### PR TITLE
Fix: wrong denom check at RegsterCoin and remove IBC-related tests

### PR DESCRIPTION
**Pull request by jasonsong0**
## Description

This is based on the  following PRs patched in canto v8. 
* [feat(erc20): add indexes at genesis state to track active token pairs Canto#68](https://github.com/b-harvest/Canto/pull/68)
* [Fix: wrong denom check at RegisterCoin Canto#78](https://github.com/b-harvest/Canto/pull/78)

### Problem

1. If multiple token pairs exist for same denom, the iteration order is based on bytes, not the registration order.

2. `coinMetadata.Base` is used as the denom for the token pair
    but `coinMetadata.Name` was used for checking registered token pair

### Solution

1. Added indexes at genesis state to track active token pair. 
2. Check registered token pair using `coinMatadata.Base`

### Proof of Work
1. [TestCase](https://github.com/jasonsong0/basechain/blob/440306a55a2b2e2eb6e79f8733ddf931f791d057/x/erc20/genesis_test.go#L170C1-L173C5): denom index should be same with lastest token pair id
2. [TestCase](https://github.com/jasonsong0/basechain/blob/ba501785e43322e41e3d5cfa39951908d93bff19/x/erc20/keeper/proposals_test.go#L221C1-L242C5): The test case for duplicate denom registering passes with the fix in the PR 

### Testcode Refactoring
- IBC-realted tests are removed because IBC features have been removed in Basechain.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] targeted the correct branch (see [PR Targeting](https://github.com/canto-network/canto/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] included the necessary unit and integration tests
- [ ] reviewed "Files changed" and left comments if necessary
 
### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by zsystm — APPROVED** [review]
LGTM

**Review by dongsam — CHANGES_REQUESTED** [review]
Need to Cherry-pick below PRs without canto v8(sdk 50) dependent code
- [ ]  https://github.com/b-harvest/Canto/pull/68
- [x]  https://github.com/b-harvest/Canto/pull/78

Could you backport patch for https://github.com/b-harvest/Canto/pull/68   or If you think a is not necessary, please leave a comment about the reason @jasonsong0

**Comment by jasonsong0** [comment]
> Need to Cherry-pick below PRs without canto v8(sdk 50) dependent code
> 
> * [ ]   [feat(erc20): add indexes at genesis state to track active token pairs Canto#68](https://github.com/b-harvest/Canto/pull/68)
> * [x]   [Fix: wrong denom check at RegisterCoin Canto#78](https://github.com/b-harvest/Canto/pull/78)
> 
> Could you backport patch for [b-harvest#68](https://github.com/b-harvest/Canto/pull/68) and Review request again? @jasonsong0

The index was added as a patch for the existence of multiple token pairs with the same denom, which should not happen if the above PR is fixed.
The name change better serves its purpose, so I'll apply the commit.

**Comment by dongsam** [comment]
@jasonsong0 It might be a good to backporting patch proceed with https://github.com/b-harvest/Canto/pull/68 even after applying the duplication prevention patch https://github.com/b-harvest/Canto/pull/78 for the following reasons:
- The genesis export/import logic has been clearly modified.
- The DenomMap(`TokenPairDenomIndex`) and ERC20Map(`TokenPairERC20AddressIndex`), which were missing in the genesis state, have been explicitly added.
- By using common code and naming through backporting, there are benefits in terms of future maintenance.

@zsystm , If you have any opinions on this as well, Please let us know

**Comment by zsystm** [comment]
> @jasonsong0 It might be a good to backporting patch proceed with [b-harvest#68](https://github.com/b-harvest/Canto/pull/68) even after applying the duplication prevention patch [b-harvest#78](https://github.com/b-harvest/Canto/pull/78) for the following reasons:
> 
> * The genesis export/import logic has been clearly modified.
> * The DenomMap(`TokenPairDenomIndex`) and ERC20Map(`TokenPairERC20AddressIndex`), which were missing in the genesis state, have been explicitly added.
> * By using common code and naming through backporting, there are benefits in terms of future maintenance.
> 
> @zsystm , If you have any opinions on this as well, Please let us know

Agree with @dongsam 
@jasonsong0 I think it's good to have https://github.com/b-harvest/Canto/pull/68

Ah ha. You already added commits. I'll review it.

File: x/erc20/genesis.go [hunk 1]
```diff
@@ -23,18 +24,31 @@ func InitGenesis(
...
 	for _, pair := range data.TokenPairs {
-		id := pair.GetID()
 		k.SetTokenPair(ctx, pair)
-		k.SetDenomMap(ctx, pair.Denom, id)
-		k.SetERC20Map(ctx, pair.GetERC20Contract(), id)
 	}
+
+	// set indexes
+	// multiple contracts at the same denom can exist,
+	// but only one which is in indexes are valid.
```
Comment (zsystm) [inline x/erc20/genesis.go]: @jasonsong0 
```suggestion
	// multiple token pairs should not be registered with the same denom, 
	// but if this happens, only the one in the index is valid.
```

File: x/erc20/genesis_test.go [hunk 2]
```diff
@@ -21,6 +22,68 @@ import (
 	"github.com/Canto-Network/Canto/v7/x/erc20/types"
 )
 
+var (
+	uqstars = "utestdenom"
+	// uqstars1 and uqstars2 have same denom
+	// uqstars1 is deployed first.
```
Comment (zsystm) [inline x/erc20/genesis_test.go:30]: @jasonsong0 
```suggestion
        // The following test codes are to prove that multiple token pairs should not be registered 
        // with the same denom, but if this happens, only the one in the index is valid.
	// uqstars1 and uqstars2 have same denom
	// uqstars1 is deployed first.
```

**Review by zsystm — COMMENTED** [review]
@jasonsong0 

- same denom and multiple token pair bug should not happen in this codebase.
- the previous comments could be interpreted as if this issue can already occur or has occurred, so I have revised the comment

**Review by zsystm — APPROVED** [review]
LGTM
FYI. @jasonsong0 test-coverage CI failed.

File: x/erc20/genesis_test.go [hunk 3]
```diff
@@ -39,7 +104,7 @@ func (suite *GenesisTestSuite) SetupTest() {
 	suite.app = app.Setup(false, feemarkettypes.DefaultGenesisState())
 	suite.ctx = suite.app.BaseApp.NewContext(false, tmproto.Header{
 		Height:          1,
-		ChainID:         "canto_9000-1",
+		ChainID:         "test-chainid",
```
Comment (dongsam) [inline x/erc20/genesis_test.go]: The chainID of Ethermint chains should be EIP155 Ethereum-compatible [Ref](https://github.com/Canto-Network/ethermint/blob/db8920e278dbf6ed03aa63212f8bdd3f83b6e68d/types/chain_id.go#L1-L50)

if you keep `canto_9000-1`,  we can refactor at a time

Comment (dongsam) [reply x/erc20/genesis_test.go]: Please let me know if it was an intentional correction or if you had any other intentions

**Review by dongsam — APPROVED** [review]

**Comment by dongsam** [comment]
LGTM, Please merge it If all the work is done @jasonsong0

**Pull request merged**